ENH: GH9746 DataFrame.unstack and Series.unstack now take fill_value …

amcpherson opened the pull request:
…kw for filling NaN when unstack results in a sparse DataFrame

closes #9746

amcpherson committed:
ENH: GH9746 DataFrame.unstack and Series.unstack now take fill_value kw for filling NaN when unstack results in a sparse DataFrame

shoyer commented:
This may interact with #9023

amcpherson commented:
I can resolve conflicts where necessary, let me know when and how to proceed.

jreback commented on pandas/core/reshape.py:
```diff
@@ -179,6 +180,10 @@ def get_new_values(self):
         if self.mask.all():
             dtype = values.dtype
             new_values = np.empty(result_shape, dtype=dtype)
+        elif self.fill_value is not None:
```
this should happen before the ``self.mask.all()`` check I think

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -12926,6 +12926,34 @@ def test_stack_unstack(self):
 
+    def test_unstack_fill(self):
+
+        # GH #9746: fill_value keyword argument for Series
```
add a test where the ``fill_value`` changes the dtype (so I think our current impl will error, so need to handle that)

amcpherson commented:
Reworked and simplified, and now just pass fill_value to _maybe_promote, which I believe is more correct.  As an effect of this I had to handle a None fill_value passed to _maybe_promote as this is the default of unstack().  As an alternative the default of unstack() could be fill_value=np.nan.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -865,6 +865,7 @@ Changes to ``Categorical.unique``
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 ``Categorical.unique`` now returns new ``Categoricals`` with ``categories`` and ``codes`` that are unique, rather than returning ``np.array`` (:issue:`10508`)
+- ``DataFrame.unstack`` and ``Series.unstack`` now take ``fill_value`` keyword to allow direct replacement of missing values when an unstack results in missing values in the resulting ``DataFrame``.  As an added benefit, specifying ``fill_value`` will preserve the data type of the original stacked data.
```
move to 0.17.1

jreback commented on pandas/core/common.py:
```diff
@@ -1094,6 +1094,15 @@ def _maybe_promote(dtype, fill_value=np.nan):
+            dtype = np.float64
+            fill_value = np.nan
+        else:
+            dtype = np.object_
```
complex should be with float
datetime/timedelta get NaT
categorical should raise

don't use subclass checks rather use the
is_float_dtype functions

jreback replied:
add some tests for promotion I think they r in test_common

amcpherson replied:
Not sure why categorical should raise.  This would seem to be a change in overall functionality of _maybe_promote as previously a categorical would not be promoted as far as I can see.

jreback replied:
i don't recall whether it ever hits this
it should fill nan (if we have coverage) or raise (if we don't have test coverage)

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -13661,6 +13661,48 @@ def test_stack_unstack(self):
+        # From a dataframe with incorrect data type for fill_value
+        result = df.unstack(fill_value=0.5)
+
+        rows = [[1, 3, 2, 4], [0.5, 5, 0.5, 6], [7, 0.5, 8, 0.5]]
```
test with datetime/timedelta as well

jreback replied:
test with some mixed dtype frames and conversions

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -13762,6 +13762,78 @@ def test_stack_unstack(self):
+        data.index = MultiIndex.from_tuples(
+            [('x', 'a'), ('x', 'b'), ('y', 'b'), ('z', 'a')])
+
+        result = data.unstack()
```
shouldn't this also test with ``fill_value``?

amcpherson replied:
Yes... its unfortunate that TestDataFrame inherits from unittest.TestCase.  It would be nice to use the nose's generator functionality to yield a set of similar tests, but that doesnt work for unittest.TestCase classes.  I may move the tests to another class in test_frame.py

jreback replied:
you can just add another class as a mix-in. nose really doesn't support parameterized tests, so just use a loop.

amcpherson replied:
I noticed other people have done the same (added mix-in classes) for specific tests where they need to use the generator style test creation for nose.  I assume this is because the main test class TestDataFrame inherits from unittest.TestCase (to allow use of self.assert* functions from unittest.TestCase), but inheriting from unittest.TestCase prevents use of generator style tests for nose.  Seems a bit of a hack.

amcpherson replied:
I dont have a better idea, that doesnt involve copying the assert* functions into util.testing.TestCase, or does something hacky with __getattr__ in util.testing.TestCase, so ill just add a mix in class.

amcpherson replied:
Went with the simplest approach, just added a few more tests as it is more clear anyway.

jreback commented:
can you rebase/update according to comments.

jreback commented on doc/source/reshaping.rst:
```diff
@@ -228,6 +228,22 @@ which level in the columns to stack:
+
+.. ipython:: python
+   
+   df3 = df.ix[[0, 1, 4, 7], [1, 2]]
```
use ``.loc``

jreback commented on doc/source/whatsnew/v0.17.1.txt:
```diff
@@ -110,6 +110,9 @@ API changes
 - ``Series.ptp`` will now ignore missing values by default (:issue:`11163`)
+- ``DataFrame.unstack`` and ``Series.unstack`` now take ``fill_value`` keyword to allow direct replacement of
+  missing values when an unstack results in missing values in the resulting ``DataFrame``.  As an added benefit,
+  specifying ``fill_value`` will preserve the data type of the original stacked data.
```
move to 0.18.0

jreback commented on pandas/core/frame.py:
```diff
@@ -3817,6 +3817,8 @@ def unstack(self, level=-1):
         ----------
         level : int, string, or list of these, default -1 (last level)
             Level(s) of index to unstack, can pass level name
+        fill_value : replace NaN with this value if the unstack produces
```
add a ``versionadded`` tag

jreback commented on doc/source/reshaping.rst:
```diff
@@ -228,6 +228,22 @@ which level in the columns to stack:
    df2.stack('exp')
    df2.stack('animal')
 
+Unstacking can result in missing values if subgroups do not have the same
```
add a versionadded tag

jreback commented on pandas/core/series.py:
```diff
@@ -1968,6 +1968,8 @@ def unstack(self, level=-1):
         ----------
         level : int, string, or list of these, default last level
             Level(s) to unstack, can pass level name
+        fill_value : replace NaN with this value if the unstack produces
```
add a version added tag

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -13854,6 +13854,88 @@ def test_stack_unstack(self):
+                             index=list('xyz'))
+        assert_frame_equal(result, expected)
+
+        result = data.unstack(fill_value='c')
```
IIRC these are converted to ``object``, so filling with a non-category works? can you validate this

jreback commented on doc/source/whatsnew/v0.17.1.txt:
```diff
@@ -110,6 +110,9 @@ API changes
+  missing values when an unstack results in missing values in the resulting ``DataFrame``.  As an added benefit,
+  specifying ``fill_value`` will preserve the data type of the original stacked data.
 
 .. _whatsnew_0171.deprecations:
```
add the pr number with an ``:issue:`` tag (IIRC this doesn't have its own dedicated issue)

jreback replied:
actually use #9746

jreback commented:
@amcpherson only a couple of doc changes. pls ping when ready & green.

thanks for the patience.....i periodically cycle thru PR's and sometimes don't make it all the way thru....(before I start again) :)

jreback commented:
can you run: ``git diff master | flake8 --diff`` and fix anything. going to shortly be enforcing PEP8.

jreback commented:
can you rebase / update according to comments

amcpherson commented:
Apologies for the delay, let me know how it looks now.

jreback commented on doc/source/reshaping.rst:
```diff
@@ -228,6 +228,24 @@ which level in the columns to stack:
+.. versionadded: 0.18.0
+
+Unstacking can result in missing values if subgroups do not have the same
+set of labels.  By default, missing values will be replaced with NaN.
```
double back-ticks around ``NaN`` (mention that for datetimelike these will be filled with ``NaT``), so maybe say the fill value for that dtype

jreback commented on pandas/tests/frame/test_reshape.py:
```diff
@@ -136,6 +136,104 @@ def test_stack_unstack(self):
+        assert_frame_equal(result, expected)
+
+        # Test unstacking with date times
+        dv = pd.date_range('2012-01-01', periods=4).values
```
can you add a test with timedeltas as well (similar to the datetimes test)

amcpherson replied:
Yes, timedeltas required some changes to _maybe_promote, good catch.  As
with datatime, cant seem to upcast to object.

On Wed, Jan 20, 2016 at 8:48 AM Jeff Reback <notifications@github.com>
wrote:

> In pandas/tests/frame/test_reshape.py
> <https://github.com/pydata/pandas/pull/10246#discussion_r50281164>:
>
> > +        result = df.unstack(fill_value=-1)
> > +        expected['A'] = expected['A'].astype(np.int16)
> > +        expected['B'] = expected['B'].astype(np.float64)
> > +        assert_frame_equal(result, expected)
> > +
> > +        # From a dataframe with incorrect data type for fill_value
> > +        result = df.unstack(fill_value=0.5)
> > +
> > +        rows = [[1, 3, 2, 4], [0.5, 5, 0.5, 6], [7, 0.5, 8, 0.5]]
> > +        expected = DataFrame(rows, index=list('xyz'), dtype=np.float)
> > +        expected.columns = MultiIndex.from_tuples(
> > +            [('A', 'a'), ('A', 'b'), ('B', 'a'), ('B', 'b')])
> > +        assert_frame_equal(result, expected)
> > +
> > +        # Test unstacking with date times
> > +        dv = pd.date_range('2012-01-01', periods=4).values
>
> can you add a test with timedeltas as well (similar to the datetimes test)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/10246/files#r50281164>.
>

jreback replied:
gr8! also add a test with ``Period``, though these are ``object`` ATM so should work w/o changes.

jreback commented:
lmk when you have a chance to update

amcpherson commented:
Some tests failing, checking now.

amcpherson commented:
Nevermind, tests were failing due to out of date .so, all looks good.

jreback commented:
@amcpherson thanks!

jreback commented on pandas/core/common.py:
```diff
@@ -1127,6 +1127,12 @@ def _maybe_promote(dtype, fill_value=np.nan):
                     fill_value = tslib.iNaT
+            elif issubclass(dtype.type, np.timedelta64):
+                try:
+                    fill_value = lib.Timedelta(fill_value).value
```
@amcpherson It hink we can eliminate this section here, and just upcast to ``np.object_`` if the fill_value cannot be coerced, (and the above about datetimes), can you create an issue for this? (and PR would be great as well!) thanks

This is some pretty old code I think
